#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
gRPC server for fortune-rule-service.
é¢ç›¸æ‰‹ç›¸è§„åˆ™åŒ¹é…å’Œå…«å­—èåˆåˆ†ææœåŠ¡
"""

from __future__ import annotations
import logging

logger = logging.getLogger(__name__)

import json
import os
import sys
from concurrent import futures

import grpc

BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
PROJECT_ROOT = BASE_DIR
sys.path.insert(0, PROJECT_ROOT)

# å¯¼å…¥ç”Ÿæˆçš„ gRPC ä»£ç 
sys.path.insert(0, os.path.join(PROJECT_ROOT, "proto", "generated"))
import fortune_rule_pb2
import fortune_rule_pb2_grpc

from services.fortune_rule.rule_engine import FortuneRuleEngine


class FortuneRuleServicer(fortune_rule_pb2_grpc.FortuneRuleServiceServicer):
    """å®ç° FortuneRuleService çš„ gRPC æœåŠ¡"""

    def __init__(self):
        self.rule_engine = FortuneRuleEngine()

    def MatchHandRules(self, request: fortune_rule_pb2.HandRuleMatchRequest, context: grpc.ServicerContext) -> fortune_rule_pb2.RuleMatchResponse:
        """æ‰‹ç›¸è§„åˆ™åŒ¹é…å’Œå…«å­—èåˆ"""
        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        logger.info(f"[{request_time}] ğŸ“¥ fortune-rule-service: æ”¶åˆ°æ‰‹ç›¸è§„åˆ™åŒ¹é…è¯·æ±‚", flush=True)
        logger.info(f"[{request_time}] ğŸ“‹ è¯·æ±‚è¯¦æƒ…:", flush=True)
        logger.info(f"  æ‰‹ç›¸ç‰¹å¾: hand_shape={request.hand_features.hand_shape if request.hand_features else 'N/A'}", flush=True)
        logger.info(f"  å…«å­—ä¿¡æ¯: use_bazi={request.bazi_info.use_bazi if request.bazi_info else False}, date={request.bazi_info.solar_date if request.bazi_info else 'N/A'}", flush=True)
        
        try:
            # è½¬æ¢æ‰‹ç›¸ç‰¹å¾
            hand_features = {}
            if request.hand_features:
                # å®‰å…¨åœ°è½¬æ¢ map å­—æ®µ
                finger_lengths = {}
                for k, v in request.hand_features.finger_lengths.items():
                    finger_lengths[str(k)] = str(v)
                
                finger_ratios = {}
                for k, v in request.hand_features.finger_ratios.items():
                    finger_ratios[str(k)] = float(v)
                
                palm_lines = {}
                for k, v in request.hand_features.palm_lines.items():
                    palm_lines[str(k)] = str(v)
                
                measurements = {}
                for k, v in request.hand_features.measurements.items():
                    measurements[str(k)] = float(v)
                
                finger_thickness = {}
                for k, v in request.hand_features.finger_thickness.items():
                    finger_thickness[str(k)] = float(v)
                
                palm_texture = {}
                for k, v in request.hand_features.palm_texture.items():
                    palm_texture[str(k)] = str(v)
                
                hand_features = {
                    "hand_shape": request.hand_features.hand_shape,
                    "hand_shape_ratio": float(request.hand_features.hand_shape_ratio),
                    "hand_shape_confidence": float(request.hand_features.hand_shape_confidence),
                    "finger_lengths": finger_lengths,
                    "finger_ratios": finger_ratios,
                    "palm_lines": palm_lines,
                    "measurements": measurements,
                    "finger_thickness": finger_thickness,
                    "palm_texture": palm_texture,
                    "special_marks": [json.loads(m) if isinstance(m, str) and m.startswith('{') else m for m in request.hand_features.special_marks],
                    "hand_orientation": request.hand_features.hand_orientation
                }
            
            # è½¬æ¢å…«å­—æ•°æ®
            bazi_data = None
            if request.bazi_data and request.bazi_data.element_counts:
                # å®‰å…¨åœ°è½¬æ¢ map å­—æ®µ
                element_counts = {}
                for k, v in request.bazi_data.element_counts.items():
                    element_counts[str(k)] = int(v)
                
                ten_gods_stats = {}
                for k, v in request.bazi_data.ten_gods_stats.items():
                    ten_gods_stats[str(k)] = int(v)
                
                bazi_data = {
                    "element_counts": element_counts,
                    "ten_gods_stats": ten_gods_stats,
                    "bazi_pillars": json.loads(request.bazi_data.bazi_pillars) if request.bazi_data.bazi_pillars else {},
                    "metadata_json": request.bazi_data.metadata_json
                }
                # å¦‚æœ metadata_json å­˜åœ¨ï¼Œè§£æå®Œæ•´æ•°æ®
                if bazi_data.get("metadata_json"):
                    try:
                        metadata = json.loads(bazi_data["metadata_json"])
                        bazi_data.update(metadata)
                    except:
                        pass
            
            # å¦‚æœæä¾›äº†å…«å­—ä¿¡æ¯ä½†è¿˜æ²¡æœ‰å…«å­—æ•°æ®ï¼Œéœ€è¦è·å–
            if request.bazi_info and request.bazi_info.use_bazi and not bazi_data:
                logger.info(f"[{request_time}] ğŸ“¥ å¼€å§‹è·å–å…«å­—æ•°æ®...", flush=True)
                logger.info(f"[{request_time}]   æ—¥æœŸ: {request.bazi_info.solar_date}, æ—¶é—´: {request.bazi_info.solar_time}, æ€§åˆ«: {request.bazi_info.gender}", flush=True)
                bazi_data = self._get_bazi_data(
                    request.bazi_info.solar_date,
                    request.bazi_info.solar_time,
                    request.bazi_info.gender
                )
                if bazi_data:
                    logger.info(f"[{request_time}] âœ… å…«å­—æ•°æ®è·å–æˆåŠŸ", flush=True)
                    logger.info(f"[{request_time}]   äº”è¡Œç»Ÿè®¡: {bazi_data.get('element_counts', {})}", flush=True)
                    logger.info(f"[{request_time}]   åç¥ç»Ÿè®¡: {bazi_data.get('ten_gods_stats', {})}", flush=True)
                else:
                    logger.info(f"[{request_time}] âš ï¸  å…«å­—æ•°æ®è·å–å¤±è´¥", flush=True)
            
            # 1. æ‰‹ç›¸è§„åˆ™åŒ¹é…
            logger.info(f"[{request_time}] ğŸ” å¼€å§‹æ‰‹ç›¸è§„åˆ™åŒ¹é…...", flush=True)
            hand_insights = self.rule_engine.match_hand_rules(hand_features)
            logger.info(f"[{request_time}] âœ… æ‰‹ç›¸è§„åˆ™åŒ¹é…å®Œæˆï¼ŒåŒ¹é…åˆ° {len(hand_insights)} æ¡è§„åˆ™", flush=True)
            
            # 2. å…«å­—èåˆåˆ†æ
            logger.info(f"[{request_time}] ğŸ” å¼€å§‹å…«å­—èåˆåˆ†æ...", flush=True)
            integrated_insights = []
            if bazi_data:
                logger.info(f"[{request_time}] ğŸ“Š å…«å­—æ•°æ®å·²è·å–ï¼Œå¼€å§‹èåˆåˆ†æ", flush=True)
                integrated_insights = self.rule_engine.integrate_with_bazi(
                    hand_features,
                    None,  # é¢ç›¸ç‰¹å¾ï¼ˆæ‰‹ç›¸åˆ†ææ—¶ä¸ºç©ºï¼‰
                    bazi_data
                )
                logger.info(f"[{request_time}] âœ… å…«å­—èåˆåˆ†æå®Œæˆï¼ŒåŒ¹é…åˆ° {len(integrated_insights)} æ¡èåˆè§„åˆ™", flush=True)
            else:
                logger.info(f"[{request_time}] âš ï¸  å…«å­—æ•°æ®ä¸ºç©ºï¼Œè·³è¿‡èåˆåˆ†æ", flush=True)
            
            # 3. ç”Ÿæˆå»ºè®®
            recommendations = self.rule_engine.generate_recommendations(
                hand_insights,
                [],  # é¢ç›¸æ´å¯Ÿï¼ˆæ‰‹ç›¸åˆ†ææ—¶ä¸ºç©ºï¼‰
                integrated_insights
            )
            
            # æ„å»ºå“åº”
            response = fortune_rule_pb2.RuleMatchResponse()
            response.success = True
            
            # å¡«å……æ‰‹ç›¸æ´å¯Ÿ
            for insight in hand_insights:
                insight_pb = fortune_rule_pb2.AnalysisInsight(
                    category=insight.get("category", ""),
                    content=insight.get("content", ""),
                    confidence=insight.get("confidence", 0.0),
                    source=insight.get("source", "hand"),
                    feature=insight.get("feature", "")
                )
                response.hand_insights.append(insight_pb)
            
            # å¡«å……èåˆæ´å¯Ÿ
            for insight in integrated_insights:
                insight_pb = fortune_rule_pb2.AnalysisInsight(
                    category=insight.get("category", ""),
                    content=insight.get("content", ""),
                    confidence=insight.get("confidence", 0.0),
                    source=insight.get("source", "integrated"),
                    feature=insight.get("feature", "")
                )
                response.integrated_insights.append(insight_pb)
            
            # å¡«å……å»ºè®®
            response.recommendations.extend(recommendations)
            
            # å¡«å……å…«å­—æ•°æ®ï¼ˆå¦‚æœè·å–äº†ï¼‰
            if bazi_data:
                bazi_data_pb = fortune_rule_pb2.BaziData()
                # å®‰å…¨åœ°å¡«å…… element_countsï¼ˆç¡®ä¿å€¼æ˜¯ intï¼‰
                element_counts = bazi_data.get("element_counts", {})
                for k, v in element_counts.items():
                    try:
                        bazi_data_pb.element_counts[str(k)] = int(v)
                    except (ValueError, TypeError):
                        pass
                # å®‰å…¨åœ°å¡«å…… ten_gods_stats
                # æ³¨æ„ï¼šten_gods_stats å¯èƒ½æ˜¯å¤æ‚ç»“æ„ï¼Œä¸æ˜¯ç®€å•çš„ map<string, int32>
                # å¦‚æœå®ƒæ˜¯å­—ç¬¦ä¸²ï¼ˆJSONï¼‰ï¼Œåˆ™è§£æï¼›å¦‚æœæ˜¯å­—å…¸ï¼Œå°è¯•è½¬æ¢
                ten_gods_stats = bazi_data.get("ten_gods_stats", {})
                if isinstance(ten_gods_stats, str):
                    try:
                        ten_gods_stats = json.loads(ten_gods_stats)
                    except:
                        ten_gods_stats = {}
                
                # å¦‚æœ ten_gods_stats æ˜¯å­—å…¸ï¼Œå°è¯•æå–ç®€å•çš„é”®å€¼å¯¹
                # ä½† protobuf å®šä¹‰æ˜¯ map<string, int32>ï¼Œæ‰€ä»¥åªèƒ½å­˜å‚¨ç®€å•çš„æ•´æ•°ç»Ÿè®¡
                # å¤æ‚ç»“æ„å­˜å‚¨åœ¨ metadata_json ä¸­
                if isinstance(ten_gods_stats, dict):
                    # åªå­˜å‚¨ç®€å•çš„æ•´æ•°ç»Ÿè®¡ï¼Œè·³è¿‡å¤æ‚ç»“æ„
                    for k, v in ten_gods_stats.items():
                        try:
                            # å¦‚æœå€¼æ˜¯æ•´æ•°ï¼Œç›´æ¥å­˜å‚¨
                            if isinstance(v, int):
                                bazi_data_pb.ten_gods_stats[str(k)] = v
                            # å¦‚æœå€¼æ˜¯å­—ç¬¦ä¸²ä¸”å¯ä»¥è½¬æ¢ä¸ºæ•´æ•°ï¼Œå­˜å‚¨
                            elif isinstance(v, str) and v.isdigit():
                                bazi_data_pb.ten_gods_stats[str(k)] = int(v)
                        except (ValueError, TypeError):
                            pass
                # å®‰å…¨åœ°å¡«å…… bazi_pillarsï¼ˆmap<string, string>ï¼‰
                bazi_pillars = bazi_data.get("bazi_pillars", {})
                if isinstance(bazi_pillars, str):
                    try:
                        bazi_pillars = json.loads(bazi_pillars)
                    except:
                        bazi_pillars = {}
                
                if isinstance(bazi_pillars, dict):
                    for k, v in bazi_pillars.items():
                        # å¦‚æœå€¼æ˜¯å­—å…¸æˆ–å…¶ä»–å¤æ‚ç»“æ„ï¼Œè½¬ä¸º JSON å­—ç¬¦ä¸²
                        if isinstance(v, (dict, list)):
                            bazi_data_pb.bazi_pillars[str(k)] = json.dumps(v, ensure_ascii=False)
                        else:
                            bazi_data_pb.bazi_pillars[str(k)] = str(v)
                
                bazi_data_pb.metadata_json = json.dumps(bazi_data, ensure_ascii=False)
                response.bazi_data.CopyFrom(bazi_data_pb)
            
            logger.info(f"[{request_time}] âœ… fortune-rule-service: æ‰‹ç›¸è§„åˆ™åŒ¹é…å®Œæˆ", flush=True)
            return response
            
        except Exception as e:
            import traceback
            error_msg = f"æ‰‹ç›¸è§„åˆ™åŒ¹é…å¤±è´¥: {str(e)}\n{traceback.format_exc()}"
            logger.info(f"[{request_time}] âŒ fortune-rule-service: é”™è¯¯ - {error_msg}", flush=True)
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"æ‰‹ç›¸è§„åˆ™åŒ¹é…å¤±è´¥: {str(e)}")
            response = fortune_rule_pb2.RuleMatchResponse()
            response.success = False
            response.error_message = str(e)
            return response

    def MatchFaceRules(self, request: fortune_rule_pb2.FaceRuleMatchRequest, context: grpc.ServicerContext) -> fortune_rule_pb2.RuleMatchResponse:
        """é¢ç›¸è§„åˆ™åŒ¹é…å’Œå…«å­—èåˆ"""
        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        logger.info(f"[{request_time}] ğŸ“¥ fortune-rule-service: æ”¶åˆ°é¢ç›¸è§„åˆ™åŒ¹é…è¯·æ±‚", flush=True)
        logger.info(f"[{request_time}] ğŸ“‹ è¯·æ±‚è¯¦æƒ…:", flush=True)
        logger.info(f"  é¢ç›¸ç‰¹å¾: san_ting={dict(request.face_features.san_ting_ratio) if request.face_features else 'N/A'}", flush=True)
        logger.info(f"  å…«å­—ä¿¡æ¯: use_bazi={request.bazi_info.use_bazi if request.bazi_info else False}, date={request.bazi_info.solar_date if request.bazi_info else 'N/A'}", flush=True)
        
        try:
            # è½¬æ¢é¢ç›¸ç‰¹å¾
            face_features = {}
            if request.face_features:
                # å®‰å…¨åœ°è½¬æ¢ map å­—æ®µ
                san_ting_ratio = {}
                for k, v in request.face_features.san_ting_ratio.items():
                    san_ting_ratio[str(k)] = float(v)
                
                facial_attributes = {}
                for k, v in request.face_features.facial_attributes.items():
                    facial_attributes[str(k)] = str(v)
                
                feature_measurements = {}
                for k, v in request.face_features.feature_measurements.items():
                    feature_measurements[str(k)] = float(v)
                
                face_features = {
                    "san_ting_ratio": san_ting_ratio,
                    "facial_attributes": facial_attributes,
                    "feature_measurements": feature_measurements,
                    "special_features": [str(f) for f in request.face_features.special_features]
                }
            
            # è½¬æ¢å…«å­—æ•°æ®
            bazi_data = None
            if request.bazi_data and request.bazi_data.element_counts:
                # å®‰å…¨åœ°è½¬æ¢ map å­—æ®µ
                element_counts = {}
                for k, v in request.bazi_data.element_counts.items():
                    element_counts[str(k)] = int(v)
                
                ten_gods_stats = {}
                for k, v in request.bazi_data.ten_gods_stats.items():
                    ten_gods_stats[str(k)] = int(v)
                
                bazi_data = {
                    "element_counts": element_counts,
                    "ten_gods_stats": ten_gods_stats,
                    "bazi_pillars": json.loads(request.bazi_data.bazi_pillars) if request.bazi_data.bazi_pillars else {},
                    "metadata_json": request.bazi_data.metadata_json
                }
                # å¦‚æœ metadata_json å­˜åœ¨ï¼Œè§£æå®Œæ•´æ•°æ®
                if bazi_data.get("metadata_json"):
                    try:
                        metadata = json.loads(bazi_data["metadata_json"])
                        bazi_data.update(metadata)
                    except:
                        pass
            
            # å¦‚æœæä¾›äº†å…«å­—ä¿¡æ¯ä½†è¿˜æ²¡æœ‰å…«å­—æ•°æ®ï¼Œéœ€è¦è·å–
            if request.bazi_info and request.bazi_info.use_bazi and not bazi_data:
                logger.info(f"[{request_time}] ğŸ“¥ å¼€å§‹è·å–å…«å­—æ•°æ®...", flush=True)
                logger.info(f"[{request_time}]   æ—¥æœŸ: {request.bazi_info.solar_date}, æ—¶é—´: {request.bazi_info.solar_time}, æ€§åˆ«: {request.bazi_info.gender}", flush=True)
                bazi_data = self._get_bazi_data(
                    request.bazi_info.solar_date,
                    request.bazi_info.solar_time,
                    request.bazi_info.gender
                )
                if bazi_data:
                    logger.info(f"[{request_time}] âœ… å…«å­—æ•°æ®è·å–æˆåŠŸ", flush=True)
                    logger.info(f"[{request_time}]   äº”è¡Œç»Ÿè®¡: {bazi_data.get('element_counts', {})}", flush=True)
                    logger.info(f"[{request_time}]   åç¥ç»Ÿè®¡: {bazi_data.get('ten_gods_stats', {})}", flush=True)
                else:
                    logger.info(f"[{request_time}] âš ï¸  å…«å­—æ•°æ®è·å–å¤±è´¥", flush=True)
            
            # 1. é¢ç›¸è§„åˆ™åŒ¹é…
            logger.info(f"[{request_time}] ğŸ” å¼€å§‹é¢ç›¸è§„åˆ™åŒ¹é…...", flush=True)
            face_insights = self.rule_engine.match_face_rules(face_features)
            logger.info(f"[{request_time}] âœ… é¢ç›¸è§„åˆ™åŒ¹é…å®Œæˆï¼ŒåŒ¹é…åˆ° {len(face_insights)} æ¡è§„åˆ™", flush=True)
            
            # 2. å…«å­—èåˆåˆ†æ
            logger.info(f"[{request_time}] ğŸ” å¼€å§‹å…«å­—èåˆåˆ†æ...", flush=True)
            integrated_insights = []
            if bazi_data:
                logger.info(f"[{request_time}] ğŸ“Š å…«å­—æ•°æ®å·²è·å–ï¼Œå¼€å§‹èåˆåˆ†æ", flush=True)
                integrated_insights = self.rule_engine.integrate_with_bazi(
                    None,  # æ‰‹ç›¸ç‰¹å¾ï¼ˆé¢ç›¸åˆ†ææ—¶ä¸ºç©ºï¼‰
                    face_features,
                    bazi_data
                )
                logger.info(f"[{request_time}] âœ… å…«å­—èåˆåˆ†æå®Œæˆï¼ŒåŒ¹é…åˆ° {len(integrated_insights)} æ¡èåˆè§„åˆ™", flush=True)
            else:
                logger.info(f"[{request_time}] âš ï¸  å…«å­—æ•°æ®ä¸ºç©ºï¼Œè·³è¿‡èåˆåˆ†æ", flush=True)
            
            # 3. ç”Ÿæˆå»ºè®®
            recommendations = self.rule_engine.generate_recommendations(
                [],  # æ‰‹ç›¸æ´å¯Ÿï¼ˆé¢ç›¸åˆ†ææ—¶ä¸ºç©ºï¼‰
                face_insights,
                integrated_insights
            )
            
            # æ„å»ºå“åº”
            response = fortune_rule_pb2.RuleMatchResponse()
            response.success = True
            
            # å¡«å……é¢ç›¸æ´å¯Ÿ
            for insight in face_insights:
                insight_pb = fortune_rule_pb2.AnalysisInsight(
                    category=insight.get("category", ""),
                    content=insight.get("content", ""),
                    confidence=insight.get("confidence", 0.0),
                    source=insight.get("source", "face"),
                    feature=insight.get("feature", "")
                )
                response.face_insights.append(insight_pb)
            
            # å¡«å……èåˆæ´å¯Ÿ
            for insight in integrated_insights:
                insight_pb = fortune_rule_pb2.AnalysisInsight(
                    category=insight.get("category", ""),
                    content=insight.get("content", ""),
                    confidence=insight.get("confidence", 0.0),
                    source=insight.get("source", "integrated"),
                    feature=insight.get("feature", "")
                )
                response.integrated_insights.append(insight_pb)
            
            # å¡«å……å»ºè®®
            response.recommendations.extend(recommendations)
            
            # å¡«å……å…«å­—æ•°æ®ï¼ˆå¦‚æœè·å–äº†ï¼‰
            if bazi_data:
                bazi_data_pb = fortune_rule_pb2.BaziData()
                # å®‰å…¨åœ°å¡«å…… element_countsï¼ˆç¡®ä¿å€¼æ˜¯ intï¼‰
                element_counts = bazi_data.get("element_counts", {})
                for k, v in element_counts.items():
                    try:
                        bazi_data_pb.element_counts[str(k)] = int(v)
                    except (ValueError, TypeError):
                        pass
                # å®‰å…¨åœ°å¡«å…… ten_gods_stats
                # æ³¨æ„ï¼šten_gods_stats å¯èƒ½æ˜¯å¤æ‚ç»“æ„ï¼Œä¸æ˜¯ç®€å•çš„ map<string, int32>
                # å¦‚æœå®ƒæ˜¯å­—ç¬¦ä¸²ï¼ˆJSONï¼‰ï¼Œåˆ™è§£æï¼›å¦‚æœæ˜¯å­—å…¸ï¼Œå°è¯•è½¬æ¢
                ten_gods_stats = bazi_data.get("ten_gods_stats", {})
                if isinstance(ten_gods_stats, str):
                    try:
                        ten_gods_stats = json.loads(ten_gods_stats)
                    except:
                        ten_gods_stats = {}
                
                # å¦‚æœ ten_gods_stats æ˜¯å­—å…¸ï¼Œå°è¯•æå–ç®€å•çš„é”®å€¼å¯¹
                # ä½† protobuf å®šä¹‰æ˜¯ map<string, int32>ï¼Œæ‰€ä»¥åªèƒ½å­˜å‚¨ç®€å•çš„æ•´æ•°ç»Ÿè®¡
                # å¤æ‚ç»“æ„å­˜å‚¨åœ¨ metadata_json ä¸­
                if isinstance(ten_gods_stats, dict):
                    # åªå­˜å‚¨ç®€å•çš„æ•´æ•°ç»Ÿè®¡ï¼Œè·³è¿‡å¤æ‚ç»“æ„
                    for k, v in ten_gods_stats.items():
                        try:
                            # å¦‚æœå€¼æ˜¯æ•´æ•°ï¼Œç›´æ¥å­˜å‚¨
                            if isinstance(v, int):
                                bazi_data_pb.ten_gods_stats[str(k)] = v
                            # å¦‚æœå€¼æ˜¯å­—ç¬¦ä¸²ä¸”å¯ä»¥è½¬æ¢ä¸ºæ•´æ•°ï¼Œå­˜å‚¨
                            elif isinstance(v, str) and v.isdigit():
                                bazi_data_pb.ten_gods_stats[str(k)] = int(v)
                        except (ValueError, TypeError):
                            pass
                # å®‰å…¨åœ°å¡«å…… bazi_pillarsï¼ˆmap<string, string>ï¼‰
                bazi_pillars = bazi_data.get("bazi_pillars", {})
                if isinstance(bazi_pillars, str):
                    try:
                        bazi_pillars = json.loads(bazi_pillars)
                    except:
                        bazi_pillars = {}
                
                if isinstance(bazi_pillars, dict):
                    for k, v in bazi_pillars.items():
                        # å¦‚æœå€¼æ˜¯å­—å…¸æˆ–å…¶ä»–å¤æ‚ç»“æ„ï¼Œè½¬ä¸º JSON å­—ç¬¦ä¸²
                        if isinstance(v, (dict, list)):
                            bazi_data_pb.bazi_pillars[str(k)] = json.dumps(v, ensure_ascii=False)
                        else:
                            bazi_data_pb.bazi_pillars[str(k)] = str(v)
                
                bazi_data_pb.metadata_json = json.dumps(bazi_data, ensure_ascii=False)
                response.bazi_data.CopyFrom(bazi_data_pb)
            
            logger.info(f"[{request_time}] âœ… fortune-rule-service: é¢ç›¸è§„åˆ™åŒ¹é…å®Œæˆ", flush=True)
            return response
            
        except Exception as e:
            import traceback
            error_msg = f"é¢ç›¸è§„åˆ™åŒ¹é…å¤±è´¥: {str(e)}\n{traceback.format_exc()}"
            logger.info(f"[{request_time}] âŒ fortune-rule-service: é”™è¯¯ - {error_msg}", flush=True)
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"é¢ç›¸è§„åˆ™åŒ¹é…å¤±è´¥: {str(e)}")
            response = fortune_rule_pb2.RuleMatchResponse()
            response.success = False
            response.error_message = str(e)
            return response

    def HealthCheck(self, request: fortune_rule_pb2.HealthCheckRequest, context: grpc.ServicerContext) -> fortune_rule_pb2.HealthCheckResponse:
        """å¥åº·æ£€æŸ¥"""
        return fortune_rule_pb2.HealthCheckResponse(status="ok")
    
    def _get_bazi_data(self, solar_date: str, solar_time: str, gender: str):
        """è·å–å…«å­—æ•°æ®"""
        try:
            # ä¼˜å…ˆä½¿ç”¨ BaziService
            try:
                from server.services.bazi_service import BaziService
                bazi_result = BaziService.calculate_bazi_full(solar_date, solar_time, gender)
                
                if bazi_result and isinstance(bazi_result, dict):
                    # å¦‚æœè¿”å›çš„æ•°æ®æœ‰ "bazi" å­—æ®µï¼Œä½¿ç”¨å®ƒï¼›å¦åˆ™ç›´æ¥ä½¿ç”¨è¿”å›çš„æ•°æ®
                    if "bazi" in bazi_result:
                        bazi_data = bazi_result["bazi"]
                    else:
                        bazi_data = bazi_result
                    
                    # ç¡®ä¿ element_counts å’Œ ten_gods_stats æ˜¯å­—å…¸ç±»å‹
                    # å¦‚æœå®ƒä»¬æ˜¯å­—ç¬¦ä¸²ï¼ˆJSONï¼‰ï¼Œåˆ™è§£æå®ƒä»¬
                    if isinstance(bazi_data.get("element_counts"), str):
                        try:
                            bazi_data["element_counts"] = json.loads(bazi_data["element_counts"])
                        except:
                            bazi_data["element_counts"] = {}
                    
                    if isinstance(bazi_data.get("ten_gods_stats"), str):
                        try:
                            bazi_data["ten_gods_stats"] = json.loads(bazi_data["ten_gods_stats"])
                        except:
                            bazi_data["ten_gods_stats"] = {}
                    
                    # ç¡®ä¿ element_counts çš„å€¼æ˜¯æ•´æ•°
                    if isinstance(bazi_data.get("element_counts"), dict):
                        element_counts = {}
                        for k, v in bazi_data["element_counts"].items():
                            try:
                                element_counts[str(k)] = int(v)
                            except (ValueError, TypeError):
                                pass
                        bazi_data["element_counts"] = element_counts
                    
                    # ten_gods_stats å¯èƒ½æ˜¯å¤æ‚ç»“æ„ï¼Œä¿æŒåŸæ ·ï¼ˆåœ¨å¡«å…… protobuf æ—¶å¤„ç†ï¼‰
                    return bazi_data
            except ImportError:
                pass
            except Exception as e:
                logger.info(f"âš ï¸  BaziService è°ƒç”¨å¤±è´¥: {e}")
            
            # é™çº§æ–¹æ¡ˆï¼šä½¿ç”¨ BaziCoreClient
            try:
                from shared.clients.bazi_core_client_grpc import BaziCoreClient
                client = BaziCoreClient()
                bazi_result = client.calculate_bazi(solar_date, solar_time, gender)
                
                # ç¡®ä¿ element_counts çš„å€¼æ˜¯æ•´æ•°
                element_counts = {}
                for k, v in bazi_result.get("element_counts", {}).items():
                    try:
                        element_counts[str(k)] = int(v)
                    except (ValueError, TypeError):
                        pass
                
                return {
                    "basic_info": bazi_result.get("basic_info", {}),
                    "bazi_pillars": bazi_result.get("bazi_pillars", {}),
                    "element_counts": element_counts,
                    "ten_gods_stats": bazi_result.get("ten_gods_stats", {}),
                    "elements": bazi_result.get("elements", {})
                }
            except Exception as e:
                logger.info(f"âš ï¸  BaziCoreClient è°ƒç”¨å¤±è´¥: {e}")
                return None
                
        except Exception as e:
            logger.info(f"âš ï¸  è·å–å…«å­—æ•°æ®å¤±è´¥: {e}")
            return None


def serve(port: int = 9007):
    """å¯åŠ¨ gRPC æœåŠ¡å™¨ï¼ˆæ”¯æŒçƒ­æ›´æ–°ï¼‰"""
    try:
        from server.hot_reload.microservice_reloader import (
            create_hot_reload_server,
            register_microservice_reloader
        )
        
        server_options = [
            ('grpc.keepalive_time_ms', 300000),
            ('grpc.keepalive_timeout_ms', 20000),
            ('grpc.keepalive_permit_without_calls', False),
            ('grpc.http2.max_pings_without_data', 2),
            ('grpc.http2.min_time_between_pings_ms', 60000),
        ]
        
        server, reloader = create_hot_reload_server(
            service_name="fortune_rule",
            module_path="services.fortune_rule.grpc_server",
            servicer_class_name="FortuneRuleServicer",
            add_servicer_to_server_func=fortune_rule_pb2_grpc.add_FortuneRuleServiceServicer_to_server,
            port=port,
            server_options=server_options,
            max_workers=10,
            check_interval=30
        )
        
        register_microservice_reloader("fortune_rule", reloader)
        reloader.start()
        
        # create_hot_reload_server å·²ç»ç»‘å®šäº†ç«¯å£ï¼Œä¸éœ€è¦å†æ¬¡ç»‘å®š
        server.start()
        logger.info(f"âœ… Fortune Rule gRPC æœåŠ¡å·²å¯åŠ¨ï¼ˆçƒ­æ›´æ–°å·²å¯ç”¨ï¼‰ï¼Œç›‘å¬ç«¯å£: {port}")
        
        try:
            server.wait_for_termination()
        except KeyboardInterrupt:
            logger.info("\n>>> æ­£åœ¨åœæ­¢æœåŠ¡...")
            reloader.stop()
            server.stop(grace=5)
            logger.info("âœ… æœåŠ¡å·²åœæ­¢")
            
    except ImportError:
        # é™çº§åˆ°ä¼ ç»Ÿæ¨¡å¼
        server_options = [
            ('grpc.keepalive_time_ms', 300000),
            ('grpc.keepalive_timeout_ms', 20000),
            ('grpc.keepalive_permit_without_calls', False),
            ('grpc.http2.max_pings_without_data', 2),
            ('grpc.http2.min_time_between_pings_ms', 60000),
        ]
        
        server = grpc.server(
            futures.ThreadPoolExecutor(max_workers=10),
            options=server_options
        )
        fortune_rule_pb2_grpc.add_FortuneRuleServiceServicer_to_server(FortuneRuleServicer(), server)
        
        listen_addr = f"[::]:{port}"
        server.add_insecure_port(listen_addr)
        
        server.start()
        logger.info(f"âœ… Fortune Rule gRPC æœåŠ¡å·²å¯åŠ¨ï¼ˆä¼ ç»Ÿæ¨¡å¼ï¼‰ï¼Œç›‘å¬ç«¯å£: {port}")
        
        try:
            server.wait_for_termination()
        except KeyboardInterrupt:
            logger.info("\n>>> æ­£åœ¨åœæ­¢æœåŠ¡...")
            server.stop(grace=5)
            logger.info("âœ… æœåŠ¡å·²åœæ­¢")


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="å¯åŠ¨ Fortune Rule gRPC æœåŠ¡")
    parser.add_argument("--port", type=int, default=9007, help="æœåŠ¡ç«¯å£ï¼ˆé»˜è®¤: 9007ï¼‰")
    args = parser.parse_args()
    serve(args.port)

